File: src/main/scala/CapiQueryString.scala
package cql

import TokenType.*
import cql.grammar.*

import java.net.{URI, URLEncoder}
import java.nio.charset.Charset
import scala.collection.mutable.Map
import scala.collection.mutable.ListBuffer

case class CapiQueryStringError(message: String) extends Error(message)

object CapiQueryString {
  def build(
      program: QueryList
  ): String =
    val (searchStrs, otherQueries) = program.exprs.partitionMap {
      case q: QueryBinary                    => Left(strFromBinary(q))
      case QueryField(key, Some(value)) => Right(s"${key.literal.getOrElse("")}=${value.literal.getOrElse("")}")
      case QueryField(key, None) =>
        throw new CapiQueryStringError(
          s"The key '+$key' needs a value after it (e.g. +$key:tone/news)"
        )
    }

    val maybeSearchStr = searchStrs match {
      case Nil => None
      case strs =>
        val encodedString =
          new URI(null, null, strs.mkString(" "), null).toASCIIString
        Some(s"q=$encodedString")
    }

    val maybeOtherQueries = otherQueries match {
      case Nil  => None
      case strs => Some(strs.mkString(""))
    }

    val queryStr = List(maybeSearchStr, maybeOtherQueries).flatten.mkString("&")

    queryStr

  private def strFromContent(queryContent: QueryContent): String =
    queryContent.content match {
      case QueryStr(str)       => str
      case QueryGroup(content) => s"(${strFromBinary(content)})"
      case q: QueryBinary      => strFromBinary(q)
    }

  private def strFromBinary(queryBinary: QueryBinary): String =
    val leftStr = strFromContent(queryBinary.left)
    val rightStr = queryBinary.right
      .map { case (op, content) =>
        s" ${op.tokenType.toString} ${strFromContent(content)}"
      }
      .getOrElse("")
    leftStr + rightStr
}

File: prosemirror-client/src/lang/capiQueryString.ts
import { err, ok, Result } from "../utils/result";
import { Query, QueryBinary, QueryContent } from "./ast";
import { getQueryFieldsFromQueryBinary } from "./utils";

class CapiQueryStringError extends Error {
  public constructor(message: string) {
    super(message);
  }
}

export const queryStrFromQueryList = (query: Query): Result<Error, string> => {
  const { content } = query;
  if (!content) {
    return ok("");
  }

  const searchStrs = strFromBinary(content);

  try {
    const otherQueries = getQueryFieldsFromQueryBinary(content).flatMap(
      (expr) => {
        switch (expr.type) {
          case "QueryField": {
            if (expr.value) {
              return [`${expr.key.literal ?? ""}=${expr.value.literal ?? ""}`];
            } else {
              throw new CapiQueryStringError(
                `The field '${expr.key.literal}' needs a value after it (e.g. '${expr.key.literal}:tone/news')`
              );
            }
          }
          default: {
            return [];
          }
        }
      }
    );

    const queryStr = searchStrs.length
      ? `q=${encodeURI(searchStrs.trim())}`
      : "";

    return ok([queryStr, otherQueries].filter(Boolean).flat().join("&"));
  } catch (e) {
    return err(e as Error);
  }
};

const strFromContent = (queryContent: QueryContent): string | undefined => {
  const { content } = queryContent;
  switch (content.type) {
    case "QueryStr":
      return content.searchExpr;
    case "QueryGroup":
      return `(${strFromBinary(content.content).trim()})`;
    case "QueryBinary":
      return strFromBinary(content);
    default:
      // Ignore fields
      return;
  }
};

const strFromBinary = (queryBinary: QueryBinary): string => {
  const leftStr = strFromContent(queryBinary.left);

  const rightStr = queryBinary.right
    ? `${queryBinary.right[0].lexeme} ${strFromBinary(queryBinary.right[1])}`
    : "";

  return (leftStr ?? "") + (rightStr ? ` ${rightStr.trim()} ` : "");
};
